### PR TITLE
Fix uniswap v2 swap decoding edge case

### LINKED ISSUE #5722: Approval event in swaps
## Problem Definition

An approval event in between of the spend and the receive event in a uniswap v2 swap, which causes wrong numbers in PnL.

## Spec

- [x] Fix the particular case reported by Sebastian.
image

- [x] Add a test for that case
- [ ] Check other swaps decoders and make sure that are no approval events can appear there in between spend and receive. 

Operating system: 
Rotki version: 1.27.1

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Fix bug in #5722

### CODE REVIEW

File: rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 from typing import Callable, Optional
 
 from rotkehlchen.accounting.structures.base import HistoryBaseEntry
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
```
Comment (LefterisJP) [inline rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py:4]: > Now, you might begin to wonder, "why doesn't a bugfix contain a test?"... A test actually does exist however, the user doesn't want to be doxxed and multiple attempts at finding transactions that match this edge case have been futile.
> 
> As a result of this, whoever will be reviewing this PR I'll share a diff that contains the test. Sorry for any incoveniences
> 
> TLDR: Closes #5722

Please use the data you have and add the test in the codebase without using VCR and hard-coding the transaction and its log events, but changing hashes and addresses.

I won't even look at a PR for a bugfix without a test.

File: rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py [hunk 2]
```diff
@@ -144,6 +146,42 @@ def _maybe_enrich_lp_tokens_transfers(
             lp_token_symbol=UNISWAP_PROTOCOL,
         )
 
+    def _fix_order_of_events_for_swaps(
+            self,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> list[HistoryBaseEntry]:
+        # look for the swap send & receive
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py:155]: Please add a docstring to the function

File: rotkehlchen/tests/unit/decoders/test_uniswapv2.py [hunk 3]
```diff
@@ -747,3 +748,196 @@ def test_uniswap_v2_remove_liquidity(database, ethereum_inquirer, eth_transactio
...
+
+@pytest.mark.parametrize('ethereum_accounts', [[ADDY_4]])
+def test_uniswap_v2_swap_events_order(
+        database,
+        ethereum_inquirer,
+        eth_transactions,
+        ethereum_accounts,
+):
+    """Check that the order of swap events are consecutive."""
+    tx_hash = '0x75b4d35b14f435c60e0fb9a82a70457a213c91ef829c034c743c66bec2dcd40f'
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_uniswapv2.py:761]: Can you state that the transaction hash here has been randomly selected to protect the transaction of the user? We will at some point look at this transaction, see that it doesn't match and wonder why

File: rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py [hunk 4]
```diff
@@ -144,6 +146,42 @@ def _maybe_enrich_lp_tokens_transfers(
...
+        send_event_idx, receive_event_idx = None, None
+        for idx, event in enumerate(decoded_events):
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND and
+                event.counterparty == CPT_UNISWAP_V2
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py:161]: if this happened with uniswap v2 can also happen with sushiswap

File: rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py [hunk 5]
```diff
@@ -157,5 +195,10 @@ def enricher_rules(self) -> list[Callable]:
             self._maybe_enrich_lp_tokens_transfers,
         ]
 
+    def post_decoding_rules(self) -> list[tuple[int, Callable]]:
+        return [
+            (0, self._fix_order_of_events_for_swaps),
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py:200]: As has been commented in other PR this will execute for all the transactions and is costly. We will refactor the post decoding rule to gain context of the transaction and not execute them always

File: rotkehlchen/tests/unit/decoders/test_uniswapv2.py [hunk 6]
```diff
@@ -747,3 +748,196 @@ def test_uniswap_v2_remove_liquidity(database, ethereum_inquirer, eth_transactio
...
+
+
+@pytest.mark.parametrize('ethereum_accounts', [[ADDY_4]])
+def test_uniswap_v2_swap_events_order(
+        database,
+        ethereum_inquirer,
+        eth_transactions,
+        ethereum_accounts,
+):
+    """Check that the order of swap events are consecutive."""
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/decoders/test_uniswapv2.py:760]: Please add a description of the test case because we will read it again and won't know why this test it here. Mention that it handles the case of an approval in between of the trade

File: rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py [hunk 7]
```diff
@@ -144,6 +146,42 @@ def _maybe_enrich_lp_tokens_transfers(
...
+        if (
+            send_event_idx is not None and
+            receive_event_idx is not None and
+            receive_event_idx != send_event_idx + 1
+        ):
+            maybe_reshuffle_events(
+                out_event=decoded_events[send_event_idx],
+                in_event=decoded_events[receive_event_idx],
+                events_list=decoded_events,
+            )
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py:181]: This looks to me that could happen in other cases too. You have the transaction, is it using a router or something special? if that is the case maybe uniswap v3 could also have this

**Review by yabirgb — COMMENTED** [review]
Some comments and also this is missing the changelog entry

File: rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py (hunk 7 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py:181]: It was via the router.

File: rotkehlchen/chain/ethereum/modules/sushiswap/decoder.py [hunk 8]
```diff
@@ -112,6 +124,11 @@ def decoding_rules(self) -> list[Callable]:
             self._maybe_decode_v2_liquidity_addition_and_removal,
         ]
 
+    def post_decoding_rules(self) -> list[tuple[int, Callable]]:
+        return [
+            (0, self._fix_order_of_events_for_swaps),
+        ]
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/sushiswap/decoder.py:130]: There is no need to add this post decoding rule for all three. You can check in the function that if the counterparty is any of the three we said make sure that the swaps are consecutive. This way a transactions will loop thrice trough this logic for no reason

File: rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py (hunk 7 above)
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py]: if is a thing of the router isn't possible to filter the transactions for which this function runs using the router address? We have `UNISWAP_ROUTERS`... isn't possible to do

```
if transaction.to in UNISWAP_ROUTERS:
    return
```

?

This will prevent the logic from executing for every transaction

File: rotkehlchen/chain/ethereum/modules/sushiswap/decoder.py [hunk 9]
```diff
@@ -104,6 +105,17 @@ def _maybe_enrich_lp_tokens_transfers(
             lp_token_symbol=SUSHISWAP_PROTOCOL,
         )
 
+    def _fix_order_of_events_for_swaps(
+            self,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
```
Comment (yabirgb) [inline rotkehlchen/chain/ethereum/modules/sushiswap/decoder.py:110]: Adding it as a comment since you forgot. This needs a changelog entry

**Review by yabirgb — COMMENTED** [review]
Few comments in case there are room for improvement

File: rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py (hunk 7 above)
Comment (prettyirrelevant) [reply rotkehlchen/chain/ethereum/modules/uniswap/v2/decoder.py:186]: `UNISWAP_ROUTERS` is for AUTO routers but this is via normal uniswap v2 routers

**Comment by yabirgb** [comment]
The failing test is due the test-caching repo branch not being up to date

File: docs/changelog.rst [hunk 10]
```diff
@@ -2,6 +2,7 @@
 Changelog
 =========
 
+* :bug:`5572` Trade events in Uniswap V2 swaps now occur in a consecutive order.
```
Comment (LefterisJP) [inline docs/changelog.rst:5]: This is not a good description as it tells nothing to the user and also implies all events were wrong.

```suggestion
* :bug:`5572` Fix a detection error that caused some uniswap v2 trade edge cases not to be accounted for properly in pnl report.
```

**Review by LefterisJP — APPROVED** [review]
LGTM
